components/movementsensor/gpsutils/cachedData_test.go

```diff
@@ -60,3 +63,197 @@ func TestReadingsSerial(t *testing.T) {
+		// Ensures correct angle and orientation of velocity vector
+
+		// All angles are compass headings
+
+		// Q1: 60 degrees
+		g.nmeaData = NmeaParser{
+			Speed:          speed,
+			CompassHeading: 60.,
+		}
+
+		expectedX := speed * math.Sqrt(3) / 2
+		expectedY := speed / 2
+
+		speed1, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, speed1.X, test.ShouldAlmostEqual, expectedX)
+		test.That(t, speed1.Y, test.ShouldAlmostEqual, expectedY)
+
+		// Q2: 150 degrees
+		g.nmeaData.CompassHeading = 150.
+
+		expectedX = speed / 2.
+		expectedY = speed * -math.Sqrt(3) / 2.
+
+		speed2, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, speed2.X, test.ShouldAlmostEqual, expectedX)
+		test.That(t, speed2.Y, test.ShouldAlmostEqual, expectedY)
+
+		// Q3: 225 degrees
+		g.nmeaData.CompassHeading = 225.
+
+		expectedX = speed * -math.Sqrt(2) / 2.
+		expectedY = speed * -math.Sqrt(2) / 2.
+
+		speed3, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, speed3.X, test.ShouldAlmostEqual, expectedX)
+		test.That(t, speed3.Y, test.ShouldAlmostEqual, expectedY)
+
+		// Q4: 310 degrees
+		g.nmeaData.CompassHeading = 310.
+
+		// Convert 310 degrees compass to standard Cartesian coordinates
+		expectedX = speed * math.Cos(140.*(math.Pi/180))
+		expectedY = speed * math.Sin(140.*(math.Pi/180))
+
+		speed4, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, speed4.X, test.ShouldAlmostEqual, expectedX)
+		test.That(t, speed4.Y, test.ShouldAlmostEqual, expectedY)
+	})
+}
+
+func TestAccuracy(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+	g := NewCachedData(&mockDataReader{}, logger)
+	g.err.Set(errors.New("test error"))
+
+	g.nmeaData = NmeaParser{
+		HDOP:           hAcc,
+		VDOP:           vAcc,
+		FixQuality:     fix,
+		CompassHeading: 90.,
+	}
+
+	acc, err := g.Accuracy(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeError, "test error")
+	test.That(t, acc.Hdop, test.ShouldEqual, hAcc)
+	test.That(t, acc.Vdop, test.ShouldEqual, vAcc)
+	test.That(t, acc.NmeaFix, test.ShouldEqual, fix)
+
+	acMap := acc.AccuracyMap
+	test.That(t, acMap["hDOP"], test.ShouldEqual, hAcc)
+	test.That(t, acMap["vDOP"], test.ShouldEqual, vAcc)
+}
+
+func TestCompassHeading(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+	g := NewCachedData(&mockDataReader{}, logger)
+
+	t.Run("no valid compass heading, so return last known", func(t *testing.T) {
+		g.lastCompassHeading.SetLastCompassHeading(90.)
+		g.nmeaData = NmeaParser{
+			CompassHeading: math.NaN(),
+		}
+
+		heading, err := g.CompassHeading(ctx, make(map[string]interface{}))
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, heading, test.ShouldEqual, 90.)
```

Comment:
seems like you are using values such as 90 and 180 multiple times, make them variables